### ISSUE TITLE
for/forall expression with arrays as elements does not work

### ISSUE THREAD

**Issue by vasslitvinov** [chapel-lang/chapel#11039]
The following code currently does not work:

```chpl
var ARR = [222,333];
var BRR = for idx in ARR do [9999];
```

The way it is implemented is that first a to-become-BRR array is created as an `[ARR.domain] [1..0] int`, then each of its elements `elem` is assigned as `elem = [9999]` . Since the domain of each `elem` is `{1..0}`, this assignment fails due to size mismatch.

Cf. in the old shape-less world, each `elem` would be **initialized** from `[9999]` , which succeeds. There, `BRR` is created **after** all its elements have been computed. The corresponding code is:

```chpl
iter ITR() { yield 222; yield 333; }
var BRR = for idx in ITR() do [9999];
// BRR.type == [1..2] [BRR[1].domain] int
```

BTW an open question is what the type `BRR.eltType` should be. It makes sense for it to be the same as `BRR[1].type`, i.e. the type of the first element of the resulting array, if all BRR's elements are of the same type. Which will be always the case if `BRR[1].type` is neither an array nor a domain, i.e. does not have a runtime-type component.

If BRR's elements are arrays, as in the above code examples, it may well be that `BRR[1].domain` is different than `BRR[2].domain`. They may be different objects or (skyline) even have different sizes. Which of those domains should be the domain of `BRR.eltType` ? The current implementation makes that an empty domain, which does not work as discussed above.

Test:

    test/arrays/vass/literal-indexing.chpl

**Comment by vasslitvinov** [comment]
I am thinking that `noinit` on arrays is not exactly what we need for this.

Here is what I think happens when we use `noinit` to create the array to become `BRR` in the main issue text above. To emphasize my concern, consider the case where `ARR.eltType` is a Block-distributed domain. For example:

```chpl
// user code
proc makeArray(elm) return (some newly-created Block-distributed array);
var ARR = [222,333];
var BRR = for elm in ARR do makeArray(elm);
```
```chpl
// potential `noinit`-based implementation
const D = {1..n} dmapped Block({1..n});
type ElmType = [D] real;
var BRR: [ARR.domain] ElmType = noinit;
for (b, elm) in zip(BRR, ARR) do
  b.init(makeArray(elm));
```

My understanding is that `noinit` on an array implicitly invokes `noinit` recursively on each of its elements. In our case, that means creating a Block-distributed array (i.e. of the type ElmType) in each element of `BRR`, initialized with `noinit`.

This has non-trivial runtime cost. Which will be wasted because that array will be overwritten by the array returned by `makeArray()`.

So instead we want, when declaring `BRR`, to leave each of its element completely uninitialized. This is how it is done in the old shape-less case.

In other words, either `noinit` on an array does not mean recursive `noinit` on its elements, or we need a separate mechanism for the above.

**Comment by vasslitvinov** [comment]
@bradcray @benharsh fyi
@mppf had some good ideas about how to implement this

**Comment by mppf** [comment]
I'm not sure I agree with your characterization of `noinit` but I'm not sure it matters since it is neither designed nor implemented at this point.

My suggestion is just to use a destroy/copy-init pattern in this case instead of assignment.

**Comment by lydia-duncan** [comment]
I assumed Vass was referring to the `noinit` array specific argument, which is different from the intended `noinit` feature (the design of which is still under construction, so far as I know).  But that might be incorrect

**Comment by vasslitvinov** [comment]
I think we want to have a way for any domain map to produce an array with uninitialized memory for its elements. I opened the question whether this would involve `noinit` (whether an array-specific argument or a feature).

Using destroy + copy-init is functionally correct. I suggest that we want to avoid it for performance reasons. This would matter particularly when the thing being destroyed+initialized is a complex data structure ex. a Block array.

**Comment by mppf** [comment]
> I think we want to have a way for any domain map to produce an array with uninitialized memory for its elements. I opened the question whether this would involve noinit (whether an array-specific argument or a feature).

That's exactly what I'd expect array noinit to do. It can't possibly initialize the elements because one of the design goals of noinit was to allow proper first-touch behavior when the user code's cores-to-elements doesn't match default initialization.
